File: Part 6 - Reinforcement Learning/Section 33 - Thompson Sampling/thompson_samplingPRACTICE.py
# -*- coding: utf-8 -*-
"""
Created on Tue Dec 24 07:50:34 2019

@author: MSI_PC
"""

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import random

dataset = pd.read_csv('Ads_CTR_Optimisation.csv')
N = 10000
d = 10
ad_selected = []
num_rewards0 = [0]*d
num_rewards1= [0]*d
for i in range(0,N):
       theta = []
       for j in range(0,d):
              theta.append(random.betavariate(num_rewards1[j]+1, num_rewards0[j]+1))
       next_index = theta.index(max(theta))
       ad_selected.append(next_index+1)
       if dataset.values[i,next_index]==1:
              num_rewards1[next_index]+=1
       else:
              num_rewards0[next_index]+=1
total_reward = sum(num_rewards1)
plt.hist(ad_selected, color='red')

File: Traditional Machine Learning Models/Reinforcement Learning/Thompson Sampling/thompson_samplingPRACTICE.R
dataset = read.csv('Ads_CTR_Optimisation.csv')
N =10000
d = 10
num_of_reward1 = integer(d)
num_of_reward0 = integer(d)
ads_selected = integer()
for (n in 1:N){
  theta = integer()
  for (i in 1:d){
    theta = append(theta, rbeta(n=1 , shape1 = num_of_reward1[i]+1, shape2 = num_of_reward0[i]+1))
  }
  next_index = match(x=max(theta), table = theta)
  ads_selected = append(ads_selected, next_index)
  if (dataset[n, next_index] == 1){
    num_of_reward1[next_index]= num_of_reward1[next_index]+1 
  }
  else{
    num_of_reward0[next_index] = num_of_reward0[next_index]+1
  }
}
total_reward = sum(num_of_reward1)
hist(ads_selected, c = 'red')